Sources/Layout/MessagesCollectionViewFlowLayout.swift

```diff
@@ -45,31 +44,51 @@ open class MessagesCollectionViewFlowLayout: UICollectionViewFlowLayout {
     open var messageLabelFont: UIFont {
         didSet {
             emojiLabelFont = messageLabelFont.withSize(2 * messageLabelFont.pointSize)
+            textMessageAttributesCache.removeAllObjects()
+            attributedTextMessageAttributesCache.removeAllObjects()
+            emojiMessageAttributesCache.removeAllObjects()
         }
     }
 
-    /// The font to be used by `TextMessageCell` for `MessageData.emoji(String)` case.
-    ///
-    /// Note: The default value of this property is 2x the `messageLabelFont`.
     internal var emojiLabelFont: UIFont
 
-    /// Determines the maximum number of `MessageCollectionViewCell` attributes to cache.
-    ///
-    /// Note: The default value of this property is 500.
-    open var attributesCacheMaxSize: Int = 500 {
-        didSet {
-            layoutContextCache.countLimit = attributesCacheMaxSize
-        }
+    // MARK: - Layout
+
+    open var textMessageLayout = TextMessageLayout() {
+        didSet { textMessageAttributesCache.removeAllObjects() }
+    }
+
+    open var attributedTextMessageLayout = TextMessageLayout() {
+        didSet { attributedTextMessageAttributesCache.removeAllObjects() }
+    }
+
+    open var emojiMessageLayout = TextMessageLayout() {
+        didSet { emojiMessageAttributesCache.removeAllObjects() }
+    }
+
+    open var photoMessageLayout = MediaMessageLayout() {
+        didSet { photoMessageAttributesCache.removeAllObjects() }
     }
 
-    typealias MessageID = NSString
-    
-    /// The cache for `MessageCellLayoutContext`.
-    /// The key is the `messageId` of the `MessageType`.
-    fileprivate var layoutContextCache = NSCache<MessageID, MessageCellLayoutContext>()
+    open var videoMessageLayout = MediaMessageLayout() {
+        didSet { videoMessageAttributesCache.removeAllObjects() }
+    }
+
+    open var locationMessageLayout = MediaMessageLayout() {
+        didSet { locationMessageAttributesCache.removeAllObjects() }
+    }
 
-    /// The `MessageCellLayoutContext` for the current cell.
-    internal var currentLayoutContext: MessageCellLayoutContext!
+    open var customMessageLayout: MessageCellLayout? {
+        didSet { customMessageAttributesCache.removeAllObjects() }
+    }
+
+    private let textMessageAttributesCache = NSCache<NSString, MessageIntermediateLayoutAttributes>()
+    private let attributedTextMessageAttributesCache = NSCache<NSString, MessageIntermediateLayoutAttributes>()
+    private let emojiMessageAttributesCache = NSCache<NSString, MessageIntermediateLayoutAttributes>()
+    private let photoMessageAttributesCache = NSCache<NSString, MessageIntermediateLayoutAttributes>()
+    private let videoMessageAttributesCache = NSCache<NSString, MessageIntermediateLayoutAttributes>()
+    private let locationMessageAttributesCache = NSCache<NSString, MessageIntermediateLayoutAttributes>()
```

Comment:
Closer but still not right. We may only be able to handle text sizes 😕

Reply:
Why we define `cache` for each `messageData`? we use the `messageID` as the key, I missed something?

Reply:
Well, this way we only need to invalidate the messages that changed. If you change the layout for a `.photo` message it shouldn't affect a `.text` message

Reply:
`.photo` and `.text` has the different `messageID`, I think it would not affect each other.

Reply:
We can't just filter out all the objects in the cache that are `.photo` messages. If cache has `.text` messages and `.video` messages. We have to remove everything. By separating them, we can invalidate all attributes for a given `MessageData` case at once.